app/src/main/java/com/simplecity/amp_library/playback/MusicService.java

```diff
@@ -1696,57 +1613,48 @@ public void makeTrackShuffleList() {
         }
     }
 
-    public void makeAlbumShuffleList() {
+    public void makeAlbumShuffleList(boolean startAtFront) {
         synchronized (this) {
-            if (albumShuffleList.size() > 0) {
-                albumShuffleList = new ArrayList<>();
-            }
-
             if (playlist == null || playlist.isEmpty()) {
                 return;
             }
 
-            Map<Album, List<Song>> trackShuffleListAlbumMap = new HashMap<>();
-
-            for (Song song : trackShuffleList) {
-                Album containingAlbum = song.getAlbum();
-                if (!trackShuffleListAlbumMap.containsKey(containingAlbum)) {
-                    trackShuffleListAlbumMap.put(containingAlbum, new ArrayList<>());
-                }
-                trackShuffleListAlbumMap.get(containingAlbum).add(song);
+            if (albumShuffleList.size() > 0) {
+                albumShuffleList = new ArrayList<>();
             }
 
-            Song currentSong = null;
-            if (playPos >= 0 && playPos < playlist.size()) {
-                currentSong = trackShuffleList.get(playPos);
-            }
+            // We are shuffling the current queue, so current song is within the list.
+            // trackShuffleList is already shuffled, so we don't need to preshuffle.
+            albumShuffleList.addAll(albumShuffleSongs(
+                    trackShuffleList.isEmpty() ? playlist : trackShuffleList,
+                    false, false));
 
-            if (currentSong != null) {
-                trackShuffleListAlbumMap.get(currentSong.getAlbum()).remove(currentSong);
-                albumShuffleList.add(currentSong);
-                albumShuffleList.addAll(trackShuffleListAlbumMap.get(currentSong.getAlbum()));
-                trackShuffleListAlbumMap.remove(currentSong.getAlbum());
+            if (startAtFront) {
+                playPos = 0;
+            } else  {
+                playPos = currentSong.getAlbum().getSongsSingle().blockingGet().indexOf(currentSong);
```

Comment:
Calling `blockingGet()` is something I really advise against here - with reactive programming, I think it's always good to assume a reactive stream could potentially run for longer than you expect - this will block the entire thread until completion. It's almost never correct to call `blocking` on RX, in my opinion.

I don't have a good alternative off the top of my head. The `makeAlbumShuffleList()` method (and the other, similar methods) might need to be refactored to return a `List` via a handler - but then we'll have to ensure any code that needs to be executed after `makeAlbumShuffleList()` is executed within the handler.

If you're having trouble refactoring this, let me know and I'll take a look myself.

There's also the possibility that `currentSong` is `null`, so we need to add a null check.